### PR TITLE
User defined DecisionModule implementation are loaded at runtime

### PR DESCRIPTION
**Pull request by willu47**
- Uses the data_layer.model_loader script to load in the DecisionModule implementation
- Config for strategies now changed to be more in line with that of a sector model (`name`, `description`, `path` and `classname` keys) and reflected in the sample_project config:
```yaml
strategies:
- name: pre-specified-planning
  description: build_nuclear
  filename: build_nuke.yml
  model_name: energy_supply
- name: agent
  description: reduce emissions
  path: planning/energyagent.py
  classname: EnergyAgent
```
- The `energyagent.py` file demonstrates:
  - calls to `data_handle.get_results()`
  - querying the intervention register
  - returning a list of decision dicts e.g. [{'name': , 'build_year': }]

### CODE REVIEW

File: src/smif/decision/decision.py [hunk 1]
```diff
@@ -110,7 +125,7 @@ def decision_loop(self):
         else:
             yield {0: [x for x in self._timesteps]}
 
-    def get_decision(self, data_handle, timestep, iteration):
+    def get_decision(self, data_handle):
```
Comment (RoaldL) [inline src/smif/decision/decision.py]: Make sure to update the docstring

File: src/smif/decision/decision.py [hunk 2]
```diff
@@ -26,13 +29,13 @@ class DecisionManager(object):
...
+    The DecisionManager collates the output of the decision algorithms and
     writes the post-decision state through a DataHandle. This allows Models to access a given
     decision state (identified uniquely by timestep and decision iteration id).
 
-    A DecisionManager passes a DataHandle down to a
-    DecisionModule, allowing the DecisionModule to access model results from previous timesteps
-    and decision iterations when making decisions.
+    The DecisionManager passes a DataHandle down to a DecisionModule,
+    allowing the DecisionModule to access model results from previous timesteps
+    and decision iterations when making decisions
```
Comment (RoaldL) [inline src/smif/decision/decision.py]: Also try and keep the docstrings updated here

File: src/smif/decision/decision.py [hunk 3]
```diff
@@ -62,20 +65,32 @@ def _set_up_decision_modules(self, modelrun_name):
...
             # Extract pre-specified planning interventions
-            if strategy['strategy'] == 'pre-specified-planning':
+            if strategy['name'] == 'pre-specified-planning':
 
                 msg = "Adding %s planned interventions to pre-specified-planning %s"
                 self.logger.info(msg, len(strategy['interventions']), index)
 
                 planned_interventions.extend(strategy['interventions'])
 
+            else:
```
Comment (RoaldL) [inline src/smif/decision/decision.py]: Maybe be explicit here that this logic deals with strategy['name'] == 'agent'

File: src/smif/sample_project/config/model_runs/energy_central.yml [hunk 4]
```diff
@@ -11,7 +11,11 @@ narratives:
   technology:
   - high_tech_dsm
 strategies:
-- strategy: pre-specified-planning
+- name: pre-specified-planning
```
Comment (RoaldL) [inline src/smif/sample_project/config/model_runs/energy_central.yml]: I would suggest using a different tag here, such as 'type'. Name suggests a description of the strategy that is defined here. While pre-specified-planning or agent is more of a strategy-type.

File: src/smif/decision/decision.py (hunk 1 above)
Comment (willu47) [reply src/smif/decision/decision.py]: Fixed

File: src/smif/decision/decision.py (hunk 2 above)
Comment (willu47) [reply src/smif/decision/decision.py]: Fixed

File: src/smif/decision/decision.py (hunk 3 above)
Comment (willu47) [reply src/smif/decision/decision.py]: The exception is only for pre-specified planning modules, all other module types (rule-based or optimisation) should be loaded from the python script.

File: src/smif/sample_project/config/model_runs/energy_central.yml (hunk 4 above)
Comment (willu47) [reply src/smif/sample_project/config/model_runs/energy_central.yml]: Good idea, changing to `type`

**Pull request merged**